artichoke-backend/src/extn/core/string/string.rb

```diff
@@ -785,4 +776,67 @@ def valid_encoding?
     # NOOP and is here for compatibility.
     true
   end
+
+  private
+
+  def tr_expand_str(str)
+    arr = []
+    str_a = str.chars
+    i = 0
+    while i < str_a.length
+      if str_a[i + 1] == '-' && !str_a[i + 2].nil?
+        r = (str_a[i]..str_a[i + 2]).to_a
+        raise ArgumentError if r.empty?
+
+        r[1] = r[0] if r.length == 1
+        r.each do |c|
+          arr.push(c)
+        end
+        i += 3
+      else
+        arr.push(str_a[i])
+        i += 1
+      end
+    end
+    arr
+  end
+
+  def tr_compose_str(container, from, to, skip_double, index_check, retrieve_value)
+    last_c = ''
+    chars.map do |c|
+      index = from.index(c)
+      container << if index_check.call(index)
+                     new_c = retrieve_value.call(to, index)
+                     res = if skip_double && new_c == last_c
+                             ''
+                           else
+                             new_c
+                           end
+                     last_c = new_c
+                     res
+                   else
+                     last_c = ''
+                     c
+                   end
+    end.join
+  end
+
+  def _tr(from_str, to_str, skip_double)
+    from_str = from_str.to_str
+    to_str = to_str.to_str
+
+    result = self.class.new
+
+    if from_str.start_with?('^') && from_str.length > 1
+      from_str = from_str[1..-1]
+      from = tr_expand_str(from_str)
+      to = tr_expand_str(to_str)
+      tr_compose_str(result, from, to, skip_double, ->(index) { index.nil? }, ->(lookup, _) { lookup.last })
+    else
+      from = tr_expand_str(from_str)
+      to = tr_expand_str(to_str)
+      tr_compose_str(result, from, to, skip_double, ->(index) { !index.nil? }, ->(lookup, index) { lookup[index] || lookup.last || '' })
+    end
+    result
+  end
```

Comment:
This is super easy to follow. I like how you've factored the implementation into separate _config_ and _execute_ steps.